File: lab1/src/main/java/Service/BaseService.java
package Service;

import Domain.BaseEntity;
import Errors.SQLErrorNoEntityFound;
import Errors.ValidationError;
import Gateway.GatewayInterface;
import Validator.ValidatorInterface;

import java.util.Vector;

public class BaseService<E extends BaseEntity, G extends GatewayInterface<E>> implements ServiceInterface<E> {

  protected G gateway;
  private ValidatorInterface<E> validator;

  protected BaseService(G gateway, ValidatorInterface<E> validator) {
    this.gateway = gateway;
    this.validator = validator;
  }
  @Override
  public E find(int id) throws SQLErrorNoEntityFound {
    return gateway.find(id);
  }

  @Override
  public void delete(E entity) {
    gateway.delete(entity);
  }

  @Override
  public E update(E entity) throws ValidationError {
    validator.Validate(entity);
    return gateway.update(entity);
  }

  @Override
  public E insert(E entity) throws ValidationError {
    validator.Validate(entity);
    return gateway.insert(entity);
  }

  @Override
  public Vector<E> findAll() {
    return gateway.findAll();
  }

  public Vector<E> findLastN(int n) {
    return gateway.findLastN(n);
  }

  public int count() {
    return gateway.findAll().size();
  }
}

File: proiect_c#/mpp_csharp/src/Service/BaseService.cs

using System.Collections.Generic;
using System.Threading.Tasks;

public class BaseService<E,G>: IService<E> where E: BaseEntity where G: IRepository<E> {

  protected readonly G repository;
  private IValidator<E> validator;

  protected BaseService(G repository, IValidator<E> validator) {
    this.repository = repository;
    this.validator = validator;
  }

  public  E Find(int id) {
    return  repository.Find(id);
  }

  public void Delete(E entity) {
    repository.Delete(entity);
  }

  public void Update(E entity) {
    validator.Validate(entity);
    repository.Update(entity);
  }

  public E Insert(E entity) {
    validator.Validate(entity);
    return repository.Insert(entity);
  }

  public List<E> FindAll() {
    return repository.FindAll();
  }

  public List<E> FindLastN(int n) {
    return repository.FindLastN(n);
  }

  public  int Count() {
    var list =  repository.FindAll();
    return list.Count;
  }
}
